### PR TITLE
atan

### PR DESCRIPTION
**Pull request by matthuszagh**
Adds 2-argument arctangent function. Parameters and result are `i32` integers for fast computation. Only the 16 MSBs of the inputs are used (16 LSBs are discarded). The x and y inputs can range from -1 to 1, which corresponds to `i32::MIN` and `i32::MAX`, respectively. The output ranges from -pi to pi, which corresponds to `i32::MIN` and `i32::MAX`, respectively.

- [godbolt](https://rust.godbolt.org/z/nahKrT)

# Related
- #206

### CODE REVIEW

File: dsp/src/iir.rs [hunk 1]
```diff
@@ -1,24 +1,14 @@
...
 
+use super::abs;
 use core::f32;
 
 // These are implemented here because core::f32 doesn't have them (yet).
 // They are naive and don't handle inf/nan.
 // `compiler-intrinsics`/llvm should have better (robust, universal, and
 // faster) implementations.
 
-fn abs<T>(x: T) -> T
```
Comment (jordens) [inline dsp/src/iir.rs:11]: I think the `copysign` etc can also be elevated like this.

File: dsp/src/trig.rs [hunk 2]
```diff
@@ -143,4 +228,40 @@ mod tests {
         assert!(max_err.0 < 1.1e-5);
         assert!(max_err.1 < 1.1e-5);
     }
+
+    #[test]
+    fn atan2_absolute_error() {
```
Comment (jordens) [inline dsp/src/trig.rs:233]: We need to make sure that we explicitly test the corner case inputs as well: min, max, -1, 0, 1.

File: dsp/src/trig.rs [hunk 3]
```diff
@@ -1,8 +1,85 @@
...
+
+    if max == 0 {
+        return 0;
+    }
+
+    let ratio = (min << 15) / max;
+
+    let mut angle = {
+        // pi/4, referenced to i16::MAX
+        const PI_4_FACTOR: i32 = 25735;
```
Comment (jordens) [inline dsp/src/trig.rs:54]: Compute it, and let the compiler do that at compile time. See `cossin()`.

File: dsp/src/trig.rs [hunk 4]
```diff
@@ -1,8 +1,85 @@
...
+        // 0.285, referenced to i16::MAX
+        const FACTOR_0285: i32 = 9339;
+        // 1/pi, referenced to u16::MAX
+        const PI_INVERTED_FACTOR: i32 = 20861;
```
Comment (jordens) [inline dsp/src/trig.rs:58]: These are all "factors". Just say what they are, e.g. `PI_OVER_4` (or `PI_4`), `ZERO_285` (or just `K`), `ONE_OVER_PI` (or `INV_PI`) and let the compiler compute them.

File: dsp/src/trig.rs [hunk 5]
```diff
@@ -1,8 +1,85 @@
...
+
+        let r1 = shift_round(ratio * PI_4_FACTOR, 15);
+        let r2 = shift_round(
+            (shift_round(ratio * FACTOR_0285, 15)) * (i16::MAX as i32 - ratio),
+            15,
+        );
+        (r1 + r2) * PI_INVERTED_FACTOR
```
Comment (jordens) [inline dsp/src/trig.rs:65]: No need to round twice and then add. This should be possible to do with fewer `shift_round()`.

Also please just simplify the entire expression first. The 1/pi can be pulled in. That will simplify it significantly.

File: dsp/src/trig.rs [hunk 6]
```diff
@@ -1,8 +1,85 @@
...
+///
+/// The angle between the x-axis and the ray to the point (x,y). The
+/// result range is from i32::MIN to i32::MAX, where i32::MIN
+/// corresponds to an angle of -pi and i32::MAX corresponds to an
+/// angle of +pi.
+pub fn atan2(y: i32, x: i32) -> i32 {
+    let y = y >> 16;
+    let x = x >> 16;
+
+    let ux = abs::<i32>(x);
```
Comment (jordens) [inline dsp/src/trig.rs:28]: Do you need to specify the type here.

File: dsp/src/trig.rs [hunk 7]
```diff
@@ -143,4 +228,40 @@ mod tests {
         assert!(max_err.0 < 1.1e-5);
         assert!(max_err.1 < 1.1e-5);
     }
+
+    #[test]
```
Comment (jordens) [inline dsp/src/trig.rs:232]: I'd like a host benchmark as well. See `dsp/benches/cossin.rs`.

File: dsp/src/trig.rs [hunk 8]
```diff
@@ -1,8 +1,85 @@
...
+/// The angle between the x-axis and the ray to the point (x,y). The
+/// result range is from i32::MIN to i32::MAX, where i32::MIN
+/// corresponds to an angle of -pi and i32::MAX corresponds to an
+/// angle of +pi.
+pub fn atan2(y: i32, x: i32) -> i32 {
+    let y = y >> 16;
+    let x = x >> 16;
+
+    let ux = abs::<i32>(x);
+    let uy = abs::<i32>(y);
```
Comment (jordens) [inline dsp/src/trig.rs:29]: I would try helping the compiler a bit. You need those three octant bits (`x < 0`, `y < 0`, `uy > ux`) later again.

File: dsp/src/trig.rs [hunk 9]
```diff
@@ -1,8 +1,85 @@
...
+    //
+    // which is taken from Rajan 2006: Efficient Approximations for
+    // the Arctangent Function.
+    let (min, max) = if ux < uy { (ux, uy) } else { (uy, ux) };
+
+    if max == 0 {
+        return 0;
+    }
+
+    let ratio = (min << 15) / max;
```
Comment (jordens) [inline dsp/src/trig.rs:50]: A word on the 15 would be helpful. I know why but it takes time to derive it.

**Review by jordens — CHANGES_REQUESTED** [review]

File: dsp/src/trig.rs [hunk 10]
```diff
@@ -1,8 +1,85 @@
...
+///
+/// * `y` - Y-axis component.
+/// * `x` - X-axis component.
+///
+/// # Returns
+///
+/// The angle between the x-axis and the ray to the point (x,y). The
+/// result range is from i32::MIN to i32::MAX, where i32::MIN
+/// corresponds to an angle of -pi and i32::MAX corresponds to an
+/// angle of +pi.
```
Comment (jordens) [inline dsp/src/trig.rs:23]: Is that really correct? `i32::MAX != -i32::MIN`.

File: dsp/src/trig.rs [hunk 11]
```diff
@@ -41,39 +48,40 @@ pub fn atan2(y: i32, x: i32) -> i32 {
     //
     // which is taken from Rajan 2006: Efficient Approximations for
     // the Arctangent Function.
-    let (min, max) = if ux < uy { (ux, uy) } else { (uy, ux) };
+    let (min, max) = if y_greater { (x, y) } else { (y, x) };
```
Comment (jordens) [inline dsp/src/trig.rs:51]: Now that you don't use `y` and `x` anymore below, just

```suggestion
    if y_greater {
        core::mem::swap(&mut x, &mut y);
     }
```

and keep using those.

File: dsp/src/trig.rs (hunk 10 above)
Comment (jordens) [reply dsp/src/trig.rs:23]: I think it's actually `i32::MIN == -pi` and at the same time also `i32::MIN == pi`. Then `i32::MAX == pi*(1 - 1/(1 << 31))`

**Review by jordens — COMMENTED** [review]
Minor nit.
Looks good!

File: dsp/src/trig.rs (hunk 2 above)
Comment (matthuszagh) [reply dsp/src/trig.rs:241]: I'll add min and max. -1,0,1 are already included because `NUM_VALS` is odd. Would you like me to make this more clear with a comment, or something else?

Comment (jordens) [reply dsp/src/trig.rs]: All good like that

**Comment by jordens** [comment]
Bors r+

File: dsp/src/trig.rs (hunk 10 above)
Comment (matthuszagh) [reply dsp/src/trig.rs]: Oh, right that makes sense. It's also worth noting that `i32::MIN` is not actually reachable. I can get `i32::MAX` with `atan2(0, i32::MIN)`. The closest results are 2147461318 and -2147461318 with `atan2(1 << 16, i32::MIN))` and `atan2(-1 << 16, i32::MIN))`.

**Comment by jordens** [comment]
Bors r+
